HIP/hip-435.md

```diff
@@ -0,0 +1,243 @@
+
+**Design Goal #1**
+Mirror nodes should not be required to download and process verbose transaction information. The sidecar transaction records will be externalized in separate record files.
+
+**Design Goal #2**
+One transaction can have 0, 1, or more than 1 sidecar record related to it located in 0, 1, or more record files.
+
+**Design Goal #3**
+It should be possible and relatively easy to add new types of sidecar transaction records.
+
+## User stories
+
+As a Hedera user, I want to have the ability to access verbose information about transactions.
+
+As an operator of a mirror node, I want the ability to download and process only the most important record-related information that is externalized as part of the main RecordFile.
+
+As an operator of a mirror node, I want the ability to download sidecar record files containing all the information about records execution that a Hedera node can provide.
+
+## Specification
+
+Define new protobuf messages to support creating record files, sidecar record files, and signature files.
+
+### Record Stream File Changes
+
+A new version (V6) of the record stream file format will be created using protobufs:
+
+```
+message RecordFileObject {
+  Transaction transaction = 1;
+  TransactionRecord record = 2;
+}
+```
+
+`Transaction` and `TransactionRecord` are already defined protobuf messages and are used by the Record Stream algorithm. No changes will be required for these messages.
+
+`RecordFileObjects` are serialized in record stream files.
+
+```
+message RecordStreamFile {
+
+  // Record stream file format version
+  fixed32 recordStreamFileVersion = 1;
+
+  // Version of HAPI that was used to serialize the file
+  SemanticVersion hapiProtoVersion = 2;
+
+  // Running Hash of all RecordFileObjects before writing this file.
+  HashObject startObjectRunningHash = 3; 
+
+  // List of all the record file objects from that period.
+  repeated RecordFileObject fileObjects = 4; 
+
+  // Running Hash of all RecordFileObjects before closing this file.
+  HashObject endObjectRunningHash = 5; 
+
+  // The block number associated with this period.
+  int64 blockNumber = 6; 
+
+  // List of the hashes of all the sidecar record files created for the same period.
+  // Allows multiple sidecar files to be linked to this RecordStreamFile
+  repeated SidecarMetadata sidecars = 7; 
+}
+```
+
+`RecordStreamFile` will be used to serialize all `RecordFileObject`s that are part of the same period into record stream files. Periods are intervals of `hedera.recordStream.logPeriod` seconds used to group `RecordFileObject`s into record stream files.
+The block structure is similar to the current  structure of record files, block number and sidecars are new fields.
+
+```
+enum SidecarType {
+  ContractStateChange = 0;
+  ContractAction = 1;
```

Comment:
Another category we probably need right away is `ContractBytecode` to contain both the initcode and runtime bytecode.